flink-ml-api/src/test/java/org/apache/flink/ml/api/core/StageTest.java

```diff
@@ -0,0 +1,345 @@
+        private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+        public final Param<Integer> extraIntParam =
+                new IntParam("extraIntParam", "Description", 20, ParamValidators.alwaysTrue());
+
+        public MyStage() {
+            ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+        }
+
+        @Override
+        public Map<Param<?>, Object> getParamMap() {
+            return paramMap;
+        }
+
+        @Override
+        public void save(String path) throws IOException {
+            ReadWriteUtils.saveMetadata(this, path);
+        }
+
+        public static MyStage load(String path) throws IOException {
+            return ReadWriteUtils.loadStageParam(path);
+        }
+    }
+
+    // A Stage subclass without the static load() method.
+    private static class MyStageWithoutLoad implements Stage<MyStage>, MyParams<MyStage> {
+        private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+        public MyStageWithoutLoad() {
+            ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+        }
+
+        @Override
+        public void save(String path) throws IOException {
+            ReadWriteUtils.saveMetadata(this, path);
+        }
+
+        @Override
+        public Map<Param<?>, Object> getParamMap() {
+            return paramMap;
+        }
+    }
+
+    // Asserts that m1 and m2 are equivalent.
+    private static void assertParamMapEquals(Map<Param<?>, Object> m1, Map<Param<?>, Object> m2) {
+        Assert.assertTrue(m1 != null && m2 != null);
+        Assert.assertEquals(m1.size(), m2.size());
+
+        for (Map.Entry<Param<?>, Object> entry : m1.entrySet()) {
+            Assert.assertTrue(m2.containsKey(entry.getKey()));
+            Object v1 = entry.getValue();
+            Object v2 = m2.get(entry.getKey());
+            if (v1.getClass().isArray() && v2.getClass().isArray()) {
+                Assert.assertArrayEquals((Object[]) v1, (Object[]) v2);
+            } else {
+                Assert.assertEquals(v1, v2);
+            }
+        }
+    }
+
+    // Saves and loads the given stage. And verifies that the loaded stage has same parameter values
+    // as the original stage.
+    private static Stage<?> validateStageReadWrite(
+            Stage<?> stage, Map<String, Object> paramOverrides) throws IOException {
+        for (Map.Entry<String, Object> entry : paramOverrides.entrySet()) {
+            Param<?> param = stage.getParam(entry.getKey());
+            ReadWriteUtils.setStageParam(stage, param, entry.getValue());
+        }
+
+        String tempDir = Files.createTempDirectory("").toString();
+        String path = Paths.get(tempDir, "test").toString();
+        stage.save(path);
+        try {
+            stage.save(path);
+            Assert.fail("Expected IOException");
+        } catch (IOException e) {
+            // This is expected.
+        }
```

Comment:
Can we have a separate test method like `testSaveMultipleTimes` for this case?

Reply:
Hmm... are you suggesting to add something like the code below? In my opinion this is a bit overkill to add a dedicated method that is only used in `validateStageSaveLoad`. The logic seems pretty simple here. I can add this if you prefer.

```
private static void validateExceptionOnSecondSave(Stage<?> stage, String path) throws IOException {
    stage.save(path);

    try {
        stage.save(path);
        Assert.fail("Expected IOException");
    } catch (IOException e) {
        // This is expected.
    }
}
```